### PR TITLE
Allow to merge if file path contains " or \

### PR DESCRIPTION
**Pull request by Ilya33**
Hello.

If a filepath in a repository contains " or \ the owner can't merge pull request with this files
because "git diff-tree" adds double quotes to that filepath.
example: filepath is ab"cd but "git diff-tree" returns "ab\"cd"

Now, when the owner click "Merge Pull Request" button the server returns 500.
This commit fix it.

Best Regards, Ilya Pavlov

### CODE REVIEW

**Comment by zeripath** [comment]
Damn! I'm sorry this was missed.

I think the correct thing to do is use the -z option on git diff-tree and change the code to handle the NUL character as a field separator rather than demunge the munging.

https://git-scm.com/docs/git-diff-tree#Documentation/git-diff-tree.txt--z

This needs checking to ensure that early gits have the option though.

I wouldn't be surprised if following this there were other issues in merge with unusual filenames especially on Windows as we do a checkout - albeit a sparse checkout - so we may be affected by these.

File: services/pull/merge.go [hunk 1]
```diff
@@ -393,7 +393,12 @@ func getDiffTree(repoPath, baseBranch, headBranch string) (string, error) {
 	out := bytes.Buffer{}
 	scanner := bufio.NewScanner(strings.NewReader(list))
 	for scanner.Scan() {
-		fmt.Fprintf(&out, "/%s\n", scanner.Text())
+		filepath := scanner.Text()
+		if strings.HasPrefix(filepath, `"`) { // the filepath contains " or \.
+			filepath = strings.TrimPrefix(filepath, `"`)
```
Comment (guillep2k) [inline services/pull/merge.go]: This doesn't seem to deal with `\`.

```
$ git diff-tree a06e7ac9fdbed88f73736b876ab63dfb680037ef aad2c7e2d2e381951cbaad3be107742f0b8fc132
:100644 100644 f1e195345db6ee7433b331c3aa333b50d46088f3 5c1b14949828006ed75a3e8858957f86a2f7e2eb M      "hello\\one\\\\two"
```

The actual file name is `helo\one\\two`

And there's also other edge cases like `"with\nnew\nlines"` or `"hello\ttabs"`.

Unless we want to print them that way, which could work.

**Comment by guillep2k** [comment]
If this helps (with and without `-z`):
```
$ git diff-tree f4d8041e0cc6255902b2f55b12eaf9c10432c3bc aad2c7e2d2e381951cbaad3be107742f0b8fc132
:100644 000000 16283e876748c734fcb345c227f69869b568aad3 0000000000000000000000000000000000000000 D      "hello\ttabs"
:100644 100644 f1e195345db6ee7433b331c3aa333b50d46088f3 5c1b14949828006ed75a3e8858957f86a2f7e2eb M      "helo\\one\\\\two"
:100644 000000 4558c9ad94739d4c00005dd28da0c04771f83a87 0000000000000000000000000000000000000000 D      "with\nnew\nlines"

$ git diff-tree -z f4d8041e0cc6255902b2f55b12eaf9c10432c3bc aad2c7e2d2e381951cbaad3be107742f0b8fc132
:100644 000000 16283e876748c734fcb345c227f69869b568aad3 0000000000000000000000000000000000000000 Dhello tabs:100644 100644 f1e195345db6ee7433b331c3aa333b50d46088f3 5c1b14949828006ed75a3e8858957f86a2f7e2eb Mhelo\one\\two:100644 000000 4558c9ad94739d4c00005dd28da0c04771f83a87 0000000000000000000000000000000000000000 Dwith
new
lines

$ git diff-tree -z f4d8041e0cc6255902b2f55b12eaf9c10432c3bc aad2c7e2d2e381951cbaad3be107742f0b8fc | xxd
0000000: 3a31 3030 3634 3420 3030 3030 3030 2031  :100644 000000 1
0000010: 3632 3833 6538 3736 3734 3863 3733 3466  6283e876748c734f
0000020: 6362 3334 3563 3232 3766 3639 3836 3962  cb345c227f69869b
0000030: 3536 3861 6164 3320 3030 3030 3030 3030  568aad3 00000000
0000040: 3030 3030 3030 3030 3030 3030 3030 3030  0000000000000000
0000050: 3030 3030 3030 3030 3030 3030 3030 3030  0000000000000000
0000060: 2044 0068 656c 6c6f 0974 6162 7300 3a31   D.hello.tabs.:1
0000070: 3030 3634 3420 3130 3036 3434 2066 3165  00644 100644 f1e
0000080: 3139 3533 3435 6462 3665 6537 3433 3362  195345db6ee7433b
0000090: 3333 3163 3361 6133 3333 6235 3064 3436  331c3aa333b50d46
00000a0: 3038 3866 3320 3563 3162 3134 3934 3938  088f3 5c1b149498
00000b0: 3238 3030 3665 6437 3561 3365 3838 3538  28006ed75a3e8858
00000c0: 3935 3766 3836 6132 6637 6532 6562 204d  957f86a2f7e2eb M
00000d0: 0068 656c 6f5c 6f6e 655c 5c74 776f 003a  .helo\one\\two.:
00000e0: 3130 3036 3434 2030 3030 3030 3020 3435  100644 000000 45
00000f0: 3538 6339 6164 3934 3733 3964 3463 3030  58c9ad94739d4c00
0000100: 3030 3564 6432 3864 6130 6330 3437 3731  005dd28da0c04771
0000110: 6638 3361 3837 2030 3030 3030 3030 3030  f83a87 000000000
0000120: 3030 3030 3030 3030 3030 3030 3030 3030  0000000000000000
0000130: 3030 3030 3030 3030 3030 3030 3030 3020  000000000000000
0000140: 4400 686f 6c61 0a71 7565 0a74 616c 00    D.hola.que.tal.
```

**Comment by lunny** [comment]
It's better to add a unit test.

**Comment by Ilya33** [comment]
>  think the correct thing to do is use the -z option on git diff-tree and change the code to handle the NUL character as a field separator rather than demunge the munging.

I will try it

**Comment by Ilya33** [comment]
- added `-z` option to `getDiffTree`
- escape right trailing spaces
- escape an optional prefix `!`
- escape `*`, `?`, `[` symbols

Don't know what to do with the `\n` symbol in the filename because it is used in `sparse-checkout` as string separator and there is no escape sequence for this one.

> It's better to add a unit test.

Should I create a new file for it or just add it to `pull_merge_test.go`?

File: services/pull/merge.go [hunk 2]
```diff
@@ -378,22 +379,55 @@ func getDiffTree(repoPath, baseBranch, headBranch string) (string, error) {
...
+		// Trailing spaces
+		filepath = rightTrailingSpacesRE.ReplaceAllStringFunc(filepath, func(s string) string {
+			return strings.Repeat(`\ `, len(s))
+		})
+
+		// An optional prefix !
+		filepath = optionalNPrefixRE.ReplaceAllString(filepath, `\$1`)
+
+		// * ? [
+		filepath = strings.Replace(filepath, "*", `\*`, -1)
```
Comment (guillep2k) [inline services/pull/merge.go]: You could use [regexp.ReplaceAllString()](https://play.golang.org/p/Iyetag494tH) for this as well. However, I'm not sure about the list of special characters: why only these three specifically (I know they are used in the glob algorithm)? You'd also need to escape at least `\t` and `\\`, and I'd escape all spaces (not only trailing).

Does git enforce UTF-8 validity of the file names? (e.g. can a file with a 0x08 exist in the repo? What about ANSI encodings?)

File: services/pull/merge.go [hunk 3]
```diff
@@ -378,22 +379,55 @@ func getDiffTree(repoPath, baseBranch, headBranch string) (string, error) {
...
+		filepath = strings.Replace(filepath, "?", `\?`, -1)
+		filepath = strings.Replace(filepath, "[", `\[`, -1)
+fmt.Printf("%s\n", filepath)
+		fmt.Fprintf(&out, "/%s\n", filepath)
```
Comment (guillep2k) [inline services/pull/merge.go:417]: These lines slipped through. 😉

**Review by guillep2k — CHANGES_REQUESTED** [review]

**Comment by guillep2k** [comment]
>
> 
> Should I create a new file for it or just add it to `pull_merge_test.go`?

It looks like `services/pull/merge_test.go` is a good place for it.

**Comment by guillep2k** [comment]
Just for kicks, I've tested and git has no problem creating and patching files with 0x08 (^H) and 0x7f (DEL) in their names. But it has some problems with wrong encodings:

```
$ ls | xxd
0000000: 6814 6c6c 6f0a 7878 7808 7575 750a 7878  h.llo.xxx.uuu.xx
0000010: 787f 7979 790a                           x.yyy.
```
First file was added last, is supposed to have a \0244 (`á` in CP-1252) in its name. Now:
```
$ git diff HEAD~1
diff --git "a/h\024llo" "b/h\024llo"
new file mode 100644
index 0000000..304c8f6
--- /dev/null
+++ "b/h\024llo"
@@ -0,0 +1 @@
+badenc
```
This is actually wrong, because `h\024llo` should be `h\0244llo`. 😆

File: services/pull/merge.go (hunk 2 above)
Comment (Ilya33) [reply services/pull/merge.go]: > why only these three specifically

git `sparse_checkout` uses the same syntax as `.gitignore` files ([link](https://git-scm.com/docs/git-read-tree#_sparse_checkout)).
`.gitignore` [pattern format](https://git-scm.com/docs/gitignore#_pattern_format).

I added `\\` to escaping symbols. `\t` works fine without escaping.

File: services/pull/merge.go (hunk 3 above)
Comment (Ilya33) [reply services/pull/merge.go:417]: my bad :sweat:

File: services/pull/merge.go [hunk 4]
```diff
@@ -374,6 +375,8 @@ func Merge(pr *models.PullRequest, doer *models.User, baseGitRepo *git.Repositor
 	return nil
 }
 
+var escapedSymbols = regexp.MustCompile(`([*[?! ])`)
```
Comment (guillep2k) [inline services/pull/merge.go]: You still can add `\` to the regexp, as long as you escape it as well:
```
var escapedSymbols = regexp.MustCompile(`([*[?! \\])`)
```

**Review by guillep2k — APPROVED** [review]

**Review by zeripath — APPROVED** [review]
This will solve the issue.  I think there are several better ways of doing this - but those are more refactors than bugfixes. (I also am suspicious that there could be a bug to do with renamed files.)

**Comment by zeripath** [comment]
Please send backports

**Pull request merged**